flink-connector-mongodb/src/main/java/org/apache/flink/connector/mongodb/source/enumerator/splitter/MongoSplitters.java

```diff
@@ -0,0 +1,110 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.mongodb.source.enumerator.splitter;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.connector.mongodb.common.config.MongoConnectionOptions;
+import org.apache.flink.connector.mongodb.common.utils.MongoUtils;
+import org.apache.flink.connector.mongodb.source.config.MongoReadOptions;
+import org.apache.flink.connector.mongodb.source.split.MongoScanSourceSplit;
+import org.apache.flink.connector.mongodb.source.split.MongoSourceSplit;
+
+import com.mongodb.MongoNamespace;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoClients;
+import org.bson.BsonDocument;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.Collection;
+
+import static org.apache.flink.connector.mongodb.common.utils.MongoConstants.ERROR_MESSAGE_FIELD;
+import static org.apache.flink.connector.mongodb.common.utils.MongoUtils.isCommandSucceed;
+
+/** To split collections of MongoDB to {@link MongoSourceSplit}s. */
+@Internal
+public class MongoSplitters implements Serializable, Closeable {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final Logger LOG = LoggerFactory.getLogger(MongoSplitters.class);
+
+    private final MongoReadOptions readOptions;
+    private final boolean limitPushedDown;
+    private final MongoClient mongoClient;
+
+    public MongoSplitters(
+            MongoConnectionOptions connectionOptions,
+            MongoReadOptions readOptions,
+            boolean limitPushedDown) {
+        this.readOptions = readOptions;
+        this.limitPushedDown = limitPushedDown;
+        this.mongoClient = MongoClients.create(connectionOptions.getUri());
+    }
+
+    public Collection<MongoScanSourceSplit> split(MongoNamespace namespace) {
+        BsonDocument collStats = MongoUtils.collStats(mongoClient, namespace);
+        if (!isCommandSucceed(collStats)) {
+            LOG.error(
+                    "Execute command collStats failed: {}",
+                    collStats.getString(ERROR_MESSAGE_FIELD));
+            throw new IllegalStateException(String.format("Collection not found %s", namespace));
+        }
+
+        MongoSplitContext splitContext =
+                MongoSplitContext.of(readOptions, mongoClient, namespace, collStats);
+
+        if (limitPushedDown) {
+            LOG.info("Limit {} is applied, using single splitter", limitPushedDown);
+            return MongoSingleSplitter.INSTANCE.split(splitContext);
+        }
```

Comment:
Do we assume that if a limit was applied the number of records that we consume is so low that splitting doesn't make sense, or is there another reason for why we don't try to split things?

Reply:
Another reason for why we don't try to split things is that it is difficult for us to merge the results of each chunk when reading concurrently. We need to cache these results to sort and limit, which may cause a lot of overhead.

Reply:
@zentol 
I made some test for that case, the `Table` runtime indeed deal with the rest but the `DataStream` runtime seems not. 
Is it necessary to allow limits under the `DataStream` API as well ?
If necessary, where would it be appropriate to deal with the rest, maybe `RecordEmitter` ?